From 52afbeda22572729b65d919e10e46dc7043ccc25 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Tue, 24 Nov 2015 02:45:06 +0200
Subject: [PATCH 0551/1240] fix: comphy: a3700: Fix the dedicated PHYs init
 procedure

- Fix detection of usb2/usb3 and sdio/emmc nodes in DT
- Remove unneeded definitions
- Add second USB2 PHY base address to the header
- TODO - add support for second UTMI (USB2) PHY
  according to JIRA SYSTEMSW-2055

Change-Id: I1e0933ad944015f2efb8e630c011df2be334734d
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/25193
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Tested-by: Nadav Haklai <nadavh@marvell.com>
(cherry picked from commit 5a979115ae7c7dc5e6071fc4c65df0284382cb4a)
Reviewed-on: http://vgitil04.il.marvell.com:8080/25363
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Tested-by: Hanna Hawa <hannah@marvell.com>
---
 drivers/phy/comphy_a3700.c | 50 ++++++++++++++++++++++++++++++----------------
 drivers/phy/comphy_a3700.h |  5 +----
 2 files changed, 34 insertions(+), 21 deletions(-)

diff --git a/drivers/phy/comphy_a3700.c b/drivers/phy/comphy_a3700.c
index 73e3e67..074cf5b 100644
--- a/drivers/phy/comphy_a3700.c
+++ b/drivers/phy/comphy_a3700.c
@@ -386,12 +386,16 @@ static int comphy_usb3_power_up(void)
   *
   * return: 1 if PLL locked (OK), 0 otherwise (FAIL)
  ***************************************************************************************************/
-static int comphy_usb2_power_up(void)
+static int comphy_usb2_power_up(u8 usb32)
 {
 	int	ret;
 
 	debug_enter();
 
+	/* TODO - add support for second USB2 UTMI - JIRA SYSTEMSW-2055 */
+	if (usb32 == 0)
+		return 1;
+
 	/*
 	 * 1. PHY pull up and disable USB2 suspend
 	 */
@@ -639,28 +643,34 @@ static int comphy_sgmii_power_up(u32 lane)
  ***************************************************************************************************/
 static void comphy_dedicated_phys_init(void)
 {
-	int node, count, ret = 1;
+	int node, count, usb32, ret = 1;
 	const void *blob = gd->fdt_blob;
 
 	debug_enter();
 
-	count = fdtdec_find_aliases_for_id(blob, "usb",
-			COMPAT_MVEBU_USB, &node, A3700_MAX_USB_CNT);
-
-	if (count > 0) {
-		if (fdtdec_get_is_enabled(blob, node)) {
-			ret = comphy_usb2_power_up();
-			if (ret == 0)
-				error("Failed to initialize USB2 PHY\n");
-			else
-				debug("USB PHY init succeed\n");
+	for (usb32 = 0; usb32 <= 1; usb32++) {
+		/* There are 2 UTMI PHYs in this SOC.
+		   One is independendent and one is paired with USB3 port (OTG) */
+		if (usb32 == 0)
+			count = fdtdec_find_aliases_for_id(blob, "usb", COMPAT_MVEBU_USB, &node, 1);
+		else
+			count = fdtdec_find_aliases_for_id(blob, "usb3", COMPAT_MVEBU_USB3, &node, 1);
+
+		if (count > 0) {
+			if (fdtdec_get_is_enabled(blob, node)) {
+				ret = comphy_usb2_power_up(usb32);
+				if (ret == 0)
+					error("Failed to initialize UTMI PHY\n");
+				else
+					debug("UTMI PHY init succeed\n");
+			} else
+				debug("USB%d node is disabled\n", usb32 == 0 ? 2 : 3);
 		} else
-			debug("USB node is disabled\n");
-	} else
-		debug("No USB node in DT\n");
+			debug("No USB%d node in DT\n", usb32 == 0 ? 2 : 3);
+	}
 
 	count = fdtdec_find_aliases_for_id(blob, "sata",
-			COMPAT_MVEBU_SATA, &node, A3700_MAX_SATA_CNT);
+			COMPAT_MVEBU_SATA, &node, 1);
 
 	if (count > 0) {
 		if (fdtdec_get_is_enabled(blob, node)) {
@@ -676,7 +686,13 @@ static void comphy_dedicated_phys_init(void)
 
 
 	count = fdtdec_find_aliases_for_id(blob, "sdio",
-		 COMPAT_MVEBU_SDIO, &node, A3700_MAX_SDIO_CNT);
+		 COMPAT_MVEBU_SDIO, &node, 1);
+
+	if (count <= 0) {
+		debug("No SDIO node in DT, looking for MMC one\n");
+		count = fdtdec_find_aliases_for_id(blob, "mmc0",
+			COMPAT_MVEBU_XENON_MMC, &node, 1);
+	}
 
 	if (count > 0) {
 		if (fdtdec_get_is_enabled(blob, node)) {
diff --git a/drivers/phy/comphy_a3700.h b/drivers/phy/comphy_a3700.h
index f16c888..9063b24 100644
--- a/drivers/phy/comphy_a3700.h
+++ b/drivers/phy/comphy_a3700.h
@@ -27,10 +27,6 @@
 #define POLL_16B_REG			1
 #define POLL_32B_REG			0
 
-#define A3700_MAX_USB_CNT		1
-#define A3700_MAX_SATA_CNT		1
-#define A3700_MAX_SDIO_CNT		1
-
 /*************************/
 /* COMPHY SB definitions */
 /************************/
@@ -69,6 +65,7 @@
 #define USB32H_BASE			(MVEBU_REGS_BASE + 0x058000) /* usb3 host */
 #define USB3PHY_BASE			(MVEBU_REGS_BASE + 0x05C000)
 #define USB2PHY_BASE			(MVEBU_REGS_BASE + 0x05D000)
+#define USB2PHY2_BASE			(MVEBU_REGS_BASE + 0x05F000)
 #define USB32_CTRL_BASE			(MVEBU_REGS_BASE + 0x05D800)
 #define USB3PHY_SHFT			2
 
-- 
1.9.1

